recursive-patterns(15) Report an error when discard pattern is used but there is a _ in scope

gafter opened the pull request:
Note that this error does not apply to the discard *designation*, only the discard *pattern*.
Since the discard designation is new in C#8, this is not a breaking change.

Report an error if a constant named '_' is used as a constant pattern.
This error has been confirmed by the LDM on 2017-11-20

Also report a warning if a type named '_' is used in an is-type expression.
This requires compat council confirmation.

Open issues in https://github.com/dotnet/csharplang/issues/1054 describe pending LDM decisions for potentialy breaking scenarios.

This PR is not built atop any other PRs.

gafter commented:
@agocke @cston Please review this addition of a couple of new diagnostics.
/cc @dotnet/roslyn-compiler

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs:
```diff
@@ -84,7 +84,17 @@ internal virtual BoundExpression BindSwitchExpressionCore(SwitchExpressionSyntax
+            HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+            this.LookupSymbolsWithFallback(lookupResult, name, arity: 0, useSiteDiagnostics: ref useSiteDiagnostics, options: LookupOptions.AllMethodsOnArityZero);
+            diagnostics.Add(node, useSiteDiagnostics);
+            if (lookupResult.IsMultiViable)
```
Looking at `BindIdentifier` for comparison, should this be `if (lookupResult.Kind != LookupResultKind.Empty)` instead?

gafter replied:
No, I don't intend for inaccessible members to affect the result. I think that test is a mistake in `BindIdentifier` because of its affect on programs like this:

``` c#
class Base
{
    static int Main() => 0;
    private static int _ = 1;
}

class Derived: Base
{
    static void M1()
    {
        _ = M2(); // error!?
    }
    static int M2() => 2;
}
```

gafter replied:
Reported as https://github.com/dotnet/roslyn/issues/25798

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs:
```diff
@@ -84,7 +84,17 @@ internal virtual BoundExpression BindSwitchExpressionCore(SwitchExpressionSyntax
         {
-            // PROTOTYPE(patterns2): give an error if there is a bindable `_` in scope.
+            // give an error if there is a bindable `_` in scope.
+            var lookupResult = LookupResult.GetInstance();
```
Do you free this instance later?

gafter replied:
No; I need to.

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs:
```diff
@@ -84,7 +84,17 @@ internal virtual BoundExpression BindSwitchExpressionCore(SwitchExpressionSyntax
+            var lookupResult = LookupResult.GetInstance();
+            var name = node.UnderscoreToken.ValueText;
+            HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+            this.LookupSymbolsWithFallback(lookupResult, name, arity: 0, useSiteDiagnostics: ref useSiteDiagnostics, options: LookupOptions.AllMethodsOnArityZero);
```
Consider removing the named argument `useSiteDiagnostics:`

gafter commented:
I believe I have responded to all comments in the new iteration. Any other comments?

@cston @agocke Can you please review this?

agocke commented on src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs:
```diff
@@ -84,7 +84,18 @@ internal virtual BoundExpression BindSwitchExpressionCore(SwitchExpressionSyntax
+            var lookupResult = LookupResult.GetInstance();
+            var name = node.UnderscoreToken.ValueText;
+            HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+            this.LookupSymbolsWithFallback(lookupResult, name, arity: 0, ref useSiteDiagnostics, options: LookupOptions.AllMethodsOnArityZero);
```
I think LookupSymbols alone would be suitable here. The doc comment for LookupSymbolsWithFallback indicates it goes through more effort in the case of failure to provide more information, but I don't think we care what happens in the case of failure.

gafter commented:
@cston Can you please review this smallish pattern-matching PR?

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:
```diff
@@ -2738,6 +2738,12 @@ private BoundExpression BindIsOperator(BinaryExpressionSyntax node, DiagnosticBa
+            if (node.Right is IdentifierNameSyntax name && name.Identifier.Text == "_")
+            {
+                // PROTOTYPE(patterns2): Need compat council confirmation that this new warning in existing code is acceptable.
+                diagnostics.Add(ErrorCode.WRN_IsTypeNamedUnderscore, name.Location, alias ?? (Symbol)targetType);
```
>alias [](start = 84, length = 5)

Please test the alias case if we're not already doing so.

gafter replied:
See `UnderscoreDeclaredAndDiscardPattern_04` in a new iteration.

---
In reply to: [178314642](https://github.com/dotnet/roslyn/pull/25779#discussion_r178314642) [](ancestors = 178314642)

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs:
```diff
@@ -84,7 +84,20 @@ internal virtual BoundExpression BindSwitchExpressionCore(SwitchExpressionSyntax
+            diagnostics.Add(node, useSiteDiagnostics);
+            if (lookupResult.IsMultiViable)
+            {
+                diagnostics.Add(ErrorCode.ERR_UnderscoreDeclaredAndDiscardPattern, node.Location, lookupResult.SingleSymbolOrDefault ?? lookupResult.Symbols[0]);
```
>lookupResult.SingleSymbolOrDefault ??  [](start = 98, length = 38)

Is this necessary?

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs:
```diff
@@ -112,6 +125,25 @@ private BoundPattern BindDiscardPattern(DiscardPatternSyntax node, TypeSymbol op
+            out bool wasExpression)
+        {
+            if (innerExpression.Kind() == SyntaxKind.IdentifierName &&
+                innerExpression is IdentifierNameSyntax name && name.Identifier.Text == "_")
```
>innerExpression is IdentifierNameSyntax name && name [](start = 16, length = 52)

`((IdentifierNameSyntax)innerExpression)`

cston commented on src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests2.cs:
```diff
@@ -1048,6 +1048,112 @@ public static void M(object o)
+                //     bool M2(object o) => o switch { 1 => true, _ => false }; // error: _ in scope
+                Diagnostic(ErrorCode.ERR_UnderscoreDeclaredAndDiscardPattern, "_").WithArguments("Program0._").WithLocation(6, 48)
+                );
+        }
```
>} [](start = 8, length = 1)

Perhaps test `o is _` when a symbol `_` is in scope that is neither a constant nor a type.

gafter replied:
Already done in `UnderscoreDeclaredAndDiscardPattern_03`.

---
In reply to: [178321208](https://github.com/dotnet/roslyn/pull/25779#discussion_r178321208) [](ancestors = 178321208)

gafter commented:
@cston I believe I have addressed all of your comments in a new iteration of this PR. Please have another look. Also let me know if you think we are now suppressing too many "cascaded" diagnostics with attribute arguments.

gafter commented:
test windows_release_unit64_prtest please
